File: src/main/java/exnihilocreatio/json/CustomColorJson.java
package exnihilocreatio.json;

import com.google.gson.*;
import exnihilocreatio.texturing.Color;
import exnihilocreatio.util.LogUtil;

import java.lang.reflect.Type;

public class CustomColorJson implements JsonDeserializer<Color>, JsonSerializer<Color> {
    @Override
    public Color deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        if (json.isJsonPrimitive()) {
            JsonPrimitive prim = json.getAsJsonPrimitive();
            if (prim.isString()) {
                return new Color(prim.getAsString());
            } else if (prim.isNumber()) {
                return new Color(prim.getAsInt());
            } else {
                LogUtil.warn("Invalid Color primitive for " + json.toString());
            }
        } else {
            JsonHelper helper = new JsonHelper(json);
            return new Color((float)helper.getDouble("r"), (float)helper.getDouble("g"), (float)helper.getDouble("b"), (float)helper.getDouble("a"));
        }

        return Color.INVALID_COLOR;
    }

    @Override
    public JsonElement serialize(Color src, Type typeOfSrc, JsonSerializationContext context) {
        return new JsonPrimitive(src.getAsHexNoAlpha());
    }
}

File: src/main/java/exnihilocreatio/json/CustomColorJson.kt
package exnihilocreatio.json

import com.google.gson.*
import exnihilocreatio.texturing.Color
import exnihilocreatio.util.LogUtil

import java.lang.reflect.Type

object CustomColorJson : JsonDeserializer<Color>, JsonSerializer<Color> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Color {
        if (json.isJsonPrimitive) {
            val prim = json.asJsonPrimitive
            when {
                prim.isString -> return Color(prim.asString)
                prim.isNumber -> return Color(prim.asInt)
                else -> LogUtil.warn("Invalid Color primitive for $json")
            }
        } else {
            val helper = JsonHelper(json)
            return Color(helper.getDouble("r").toFloat(), helper.getDouble("g").toFloat(), helper.getDouble("b").toFloat(), helper.getDouble("a").toFloat())
        }

        return Color.INVALID_COLOR
    }

    override fun serialize(src: Color, typeOfSrc: Type, context: JsonSerializationContext) =
            JsonPrimitive(src.asHexNoAlpha)

}
